### ISSUE TITLE
ORDER BY and LIMIT in WITH RECURSIVE query mis-applied

### ISSUE THREAD

**Issue by AMWJ** [kysely-org/kysely#646]
The ORDER BY and LIMIT on the first ("base case" half) of a WITH RECURSIVE query are being treated like they are on the second ("recursive" half). This is occurring in this query:

```
          const distinctValues = db
            .withRecursive("t(col)", (eb) => {
              return eb
                .selectFrom("tbl")
                .orderBy("col")
                .limit(1)
                .select("col")
                .unionAll(
                  eb
                    .selectFrom("t")
                    .select(({ selectFrom }) => selectFrom("tbl").select("col").whereRef("col", ">", "t.col").orderBy("col").limit(1).as("col"))
                    .where("col", "is not", null),
                );
            })
            .selectFrom("t")
            .select("col")
            .where("col", "is not", null);
          console.log(distinctValues.compile().sql);
```

After formatting, this outputs:

```
with recursive "t"("col") as (
  select "col" from "external_analytics"."tbl"
  union all
  select (select "col" from "external_analytics"."tbl" where "col" > "t"."col" order by "col" limit $1) as "col"
  from "t" where "col" is not null *order by "col" limit $2*)
select "col" from "t" where "col" is not null
```

As you can see, the "order by "col" limit $2" is in the wrong place in the compiled query. (This has the affect of Postgres complaining, since `ORDER BY in a recursive query is not implemented`). (For reference, the query I'm imitating comes from https://wiki.postgresql.org/wiki/Loose_indexscan#If_col_IS_NOT_NULL)

**Comment by koskimas** [comment]
The query is correct. Where else would the order by be?

**Comment by koskimas** [comment]
You could do this:

```ts
const distinctValues = db
  .withRecursive("t(col)", (eb) => {
    const q1 = eb
      .selectFrom("tbl")
      .orderBy("col")
      .limit(1)
      .select("col")
      
    const q2 = eb
      .selectFrom("t")
      .select(({ selectFrom }) => selectFrom("tbl")
        .select("col")
        .whereRef("col", ">", "t.col")
        .orderBy("col")
        .limit(1)
        .as("col")
      )
      .where("col", "is not", null)
  
    return sql<{ col: string }>`${q1} union all ${q2}`
  })
  .selectFrom("t")
  .select("col")
  .where("col", "is not", null);
```

Add parentheses where needed. I'm not sure which ones will get automatically added in the case.

**Comment by koskimas** [comment]
Actually seems about right?

https://kyse.link/?p=s&i=MgBZfwHg07DdbkG4ukPR

**Comment by AMWJ** [comment]
> The SQL is correct (not the one you pasted, that's not the result of the query above). The link you provided says "parentheses required" and Kysely doesn't add those automatically for the first query in a union. The query is not wrong. It's just not what you expect.

It's hard for me to understand how this could be not wrong. It's true that I should add parentheses for Postgres to understand my query the way I'd like it to, but how does moving the ORDER BY and LIMIT to the second query of the UNION match the value of "_The API should be as close to writing raw SQL as possible while still providing type safety._" This is a case that violates type safety (allowing the representation of a query that is not allowed), and is very far from writing raw SQL (placing the ORDER BY and LIMIT on entirely different parts of the query than they were placed on).

As well, to the parentheses point, the `eb` parameter doesn't have a `paren` method, so there doesn't appear to be a way to do this right with Kysely. It looks like the best way to do this is with a raw `sql` string, as you've done above.

Aside: Is there documentation on when Kysely will place parentheses, and when it will not? It seems from your example that it will place them automatically when interpolated into an `sql` string, but will it always do that?

**Comment by igalklebanov** [comment]
Hey 👋 

> As well, to the parentheses point, the eb parameter doesn't have a paren method, so there doesn't appear to be a way to do this right with Kysely. It looks like the best way to do this is with a raw sql string, as you've done above.

What version are you on?
https://kyse.link/?p=s&i=2Ee8HfFcDlNGa8GDS1go

**Comment by AMWJ** [comment]
@igalklebanov I appreciate that the sarcastic tone shows your frustration that this conversation continues, but the `eb` parameter in your example is different than the one I'm speaking of that lacks the `parens` method. See here: https://kyse.link/?p=s&i=42cQtE8z5WJcY0RLvK53. I've preserved all sarcasm, but added no more.

There are two `eb` parameters in this example. One (the second) is unmodified from the one you defined, has the type `ExpressionBuilder`, and has the `parens` method.

The other (the first one) has the type `QueryCreator`, and does _not_ have a `parens` method. This is the `eb` I was referring to, in the same place as my original code example. It's also the `eb` that, if it had a `parens` attribute, would allow me to write the query I linked to without resorting to writing raw sql, in the way @koskimas has.

 I am on Kysely 0.26.1, which I believe is the most current version.
